### ISSUE TITLE
Allow anonymous posting of comments

### ISSUE THREAD

**Issue by abhiabhi94** [Radi85/Comment#33]
Hey Radi.

By any chance, would you want to allow anonymous posting of comments in the application?

In case you are interested, I would suggest the possible workflow:

- Ask anonymous users to give their email alongside the comment.
- We would send an email to them with a link to verify.
- The comment would only hit the database when they verify.

Obviously, this feature would be a choice based one for a developer.

**Comment by Radi85** [comment]
Hey @abhiabhi94,

so do you think that lazy users such as me 😄 would go with this process just to post a comment?

However, I would like to add this feature after the next release and keep it optional as I believe there are some developers who may like to have it.

**Comment by abhiabhi94** [comment]
> so do you think that lazy users such as me smile would go with this process just to post a comment?

In my opinion, this would generally be used for a one-time case, when they don't really want to create an account. Although, it may be used for regular anonymous conversations.

> However, I would like to add this feature after the next release and keep it optional as I believe there are some developers who may like to have it.

Are you making some changes write now? If that's the case, I will start working on this after you finish with your changes.

**Comment by Radi85** [comment]
Yes I am currently working on new feature and I have generated a migration file that's why I want to start with this after the release.
In fact the release will be done hopefully in the next few hours.

**Comment by abhiabhi94** [comment]
Okay, no issues.

I was also thinking of adding a permalink option to every comment. If you are okay with it, I will make another issue and work on this during this weekend.

**Comment by abhiabhi94** [comment]
Hey Radi, I see that you have made a new branch staging. Currently, the master branch is behind staging. Which one am I supposed to work?

I will first work on the permalink one.

Also, I have a couple of queries before I start working on this issue.

- What should we use in the model for comment author? Django has an Anonymous user that can be useful.
- How should the process work?
    - An example would be to take the user's email and send them a unique link(using some sort of token). When they click on the link, the comment is verified, the screen shows your comment has been verified. It then maybe redirects to their comment.

**Comment by Radi85** [comment]
> Hey Radi, I see that you have made a new branch staging. Currently, the master branch is behind staging. Which one am I supposed to work?

It's not. Both has the same commit. However, please keep using master since I am using staging for testing deployment. 

> * What should we use in the model for comment author? Django has an Anonymous user that can be useful.

Anonymous user cannot help here, we can change the user field to:
```python
user = models.ForeignKey(User, blank=True, null=True)
email = models.ForeignKey(Email, blank=True, null=True)
```

We need somehow to verify that one of those two is provided when saving the model.

>   * An example would be to take the user's email and send them a unique link(using some sort of token). When they click on the link, the comment is verified, the screen shows your comment has been verified. It then maybe redirects to their comment.

I agree. For this we need to add Email model:
```python
class Email(models.Model):
    email = models.EmailField(max_length=254)
    is_verified = models.BooleanField()
```
I am thinking to make a separate model for the email and use it as foreignKey in the comment, this will help us to verify the email only once, next time the email can be directly checked if it was verified without needing this long process.

BTW, I am currently working on adding our settings module so all our settings with default values will be in one place. I will push this today.

**Comment by abhiabhi94** [comment]
>  I am thinking to make a separate model for the email and use it as foreignKey in the comment, this will help us to verify the email only once, next time the email can be directly checked if it was verified without needing this long process.

This won't really be a great idea. Suppose your email is verified and I use it to post a comment. I think the verification process has to be done each time.

**Comment by Radi85** [comment]
> This won't really be a great idea. Suppose your email is verified and I use it to post a comment. I think the verification process has to be done each time.

You are absolutely right. My concern is that the lazy user won't go through this proccess each time posting a comment, what if we add a secret code to the email model and we send it with the email verification ONE TIME only, so the user can post further comments with the email and secret without needing to verify their email for each comment.

**Comment by Radi85** [comment]
BTW, I am currently working on adding Gravatar support, so it can be used directly instead of the profile model and for anonymous user in case the profile model is still used.

**Comment by abhiabhi94** [comment]
Instead of doing this the user can obviously log in they want to post
comments regularly. My concern was mainly to cover one time comments.
Using the same secret token over different comments would raise
security concerns as well.

**Comment by abhiabhi94** [comment]
no issues. It would be nice to add an issue when you're working on it. I
guess it will help me and others who might contribute in future to
cordinate better.

**Comment by Radi85** [comment]
> Instead of doing this the user can obviously log in they want to post comments regularly. My concern was mainly to cover one time comments. Using the same secret token over different comments would raise security concerns as well.

Then I am ok with your adoption.

**Comment by Radi85** [comment]
> no issues. It would be nice to add an issue when you're working on it. I guess it will help me and others who might contribute in future to cordinate better.

well, I just found out that adding gravatar shall be blocked by this. so I will proceed after you submit the updated comment model.

**Comment by abhiabhi94** [comment]
Hey Radi, I was a bit busy due to some personal stuff. I will try to make a pull request for permalink today if I can.

It might take a day or two to do away with the implementation of this issue. You do raise a valid point to allow constant anonymous conversations.

One of the ways I think we can handle this is store a cookie in the browser (for applications using django templates). This cookie would exist for a period of two or three hours. Ideally, we would want to keep this cookie for a few hours after the last comment(this would require resetting the time period for the cookie after every comment). The cookie would be destroyed after this time period. Also this might raise security issues.

This is just a raw idea that I had at this moment. The implementation might raise a few more hurdles. 

Maybe we can first of all integrate the basic version of anonymous comments that requires validation for each comment. The continuous conversation thing may be implemented in a subsequent release.

**Comment by Radi85** [comment]
> Hey Radi, I was a bit busy due to some personal stuff. I will try to make a pull request for permalink today if I can.
> It might take a day or two to do away with the implementation of this issue. You do raise a valid point to allow constant anonymous conversations.

No worries. take your time.

> Maybe we can first of all integrate the basic version of anonymous comments that requires validation for each comment. The continuous conversation thing may be implemented in a subsequent release.

This is the best for now. There are few things need also to be handeled, e.g. the comment posted by anontmous won't be editable. So as you said proceed please with the basic concept and it can be extended in the future.

**Comment by abhiabhi94** [comment]
Hey @Radi85, I'm really sorry I couldn't reach out to you earlier. I had some personal issues due to which I couldn't give much time to work.

I am just about to start working on this right now. If everything goes well, I will try to make a pull request as soon as I can.

**Comment by Radi85** [comment]
Hey @abhiabhi94,

No worries. Hope everything goes well again over there. I actually had no time as well in the past 2 weeks to work here, I hopefully will get back here by the next week.
